Prepare for splitting the proto package

tamird opened the pull request:
This builds on #1317 but I suppose it doesn't strictly have to. I did have to modify protoc-gen-gogo a bit.

tamird committed:
Compile all the protos in one place.

This commit discards all the proto Makefiles in favour of one big
daddy proto Makefile that gets included into our main Makefile.

This uses gogo/protobuf#67.

tamird committed:
Regenerate the world

petermattis commented on mk/protobuf.mk:
```diff
@@ -0,0 +1,83 @@
+# Copyright 2014 The Cockroach Authors.
```
Let's place this in the `build` directory where other build related stuff lives.

tamird replied:
done

petermattis commented on mk/protobuf.mk:
```diff
@@ -0,0 +1,83 @@
+GOTOOLS := .go_tools
+CPPTOOLS := .cpp_tools
+
+GO_PROTOS  := $(sort $(shell find . -name *.proto -type f))
```
You should probably avoid searching in the `.git` directory. Might be better to use `git ls-tree`, though that wouldn't find files that haven't been added to git.

tamird replied:
done with `find . -not -path '*/.*' -name *.proto -type f`

petermattis commented on mk/protobuf.mk:
```diff
@@ -0,0 +1,83 @@
+	# into the storage/engine directory.
+	(cd $(ENGINE_ROOT) && ln -sf proto/*.pb.cc .)
+	# protoc thinks our c++ code needs the gogo proto file, but we only use it
+	# for compiling go with advanced options. Manuall whack the generated headers
```
`s/Manuall/Manually/g`

tamird replied:
done

petermattis commented on storage/engine/db.cc:
```diff
@@ -27,9 +27,9 @@
-#include "cockroach/proto/internal.pb.h"
+#include "proto/api.pb.h"
+#include "proto/data.pb.h"
+#include "proto/internal.pb.h"
```
What caused the includes to change? Not a big deal, but I prefer the `cockroach/` prefix.

tamird replied:
Remember I mentioned having to `cd` out of the repo? that's what gave the `cockroach/` prefix. Now that I'm not doing that, these import paths mirror the on-disk layout of the protos relative to the repository root.

petermattis replied:
I'm a bit confused. It looks to me like the files are still stored under `cockroach/proto`. How is this working?

tamird replied:
turns out it isnt!

petermattis commented on protoc-gen-gogoroach/main.go:
```diff
@@ -0,0 +1,41 @@
+	req := command.Read()
+	files := req.GetProtoFile()
+
+	vanity.ForEachFile(files, vanity.TurnOnMarshalerAll)
```
Is there any other reason for having our own protoc-gen than enabling these options? I'm not at all a fan of removing the options from the .proto files so that they can be centralized here. Doing so does nothing for the clarity of the code while making it marginally easier to add or change an option in the future which is a very rare operation.

tamird replied:
Removed - we'll discuss separately in #1317.

petermattis commented on mk/protobuf.mk:
```diff
@@ -0,0 +1,83 @@
+REPO_ROOT      := .
+ORG_ROOT       := $(REPO_ROOT)/..
+GITHUB_ROOT    := $(ORG_ROOT)/..
+GOGOPROTO_ROOT := $(GITHUB_ROOT)/gogo/protobuf
```
`GOGOPROTO` vs `GOGO_PROTO`. I don't have a preference, but let's be consistent.

tamird replied:
done

petermattis commented on mk/protobuf.mk:
```diff
@@ -0,0 +1,83 @@
+CPP_HEADERS := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.h))
+CPP_SOURCES := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.cc))
+
+$(PROTOC_PLUGIN): $(REPO_ROOT)/protoc-gen-gogoroach/main.go $(shell find $(GOGOPROTO_ROOT))
```
What does making this target dependent on the directory do?

tamird replied:
invalidates the target when files in the directory are unlinked.

petermattis replied:
Or when files in the directory are added. But don't you want to rebuild when any of the files in the directory change? I think you're reproducing logic that is already inside the `go install` command, but not doing the full checks.

tamird replied:
right, that's why this uses `find` - it depends on all the files in the directory and the directory itself.

petermattis commented on mk/protobuf.mk:
```diff
@@ -0,0 +1,83 @@
+CPP_SOURCES := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.cc))
+
+$(PROTOC_PLUGIN): $(REPO_ROOT)/protoc-gen-gogoroach/main.go $(shell find $(GOGOPROTO_ROOT))
+	go install github.com/cockroachdb/cockroach/protoc-gen-gogoroach
```
`go install` is intended to be super-fast when nothing needs to be done. Rather than trying to only rebuild the plugin when necessary, it would be simpler to just blindly `go install` it before using it.

tamird replied:
meh, this was pretty easy to do and keeps this centralized behind $(PROTOC_PLUGIN)

petermattis replied:
I don't have `protoc-gen-gogoroach`, but here is the timing for `go install` of `protoc-gen-gogo`:

```
time go install github.com/gogo/protobuf/protoc-gen-gogo

real	0m0.052s
user	0m0.031s
sys	0m0.013s
```

tamird replied:
Moot, this change has moved back to #1317

petermattis commented on mk/protobuf.mk:
```diff
@@ -0,0 +1,83 @@
+
+$(CPPTOOLS): $(PROTOC) $(GOGO_PROTO)
+	rm -f $(CPP_HEADERS) $(CPP_SOURCES)
+	touch $(CPPTOOLS)
```
You seem to be jumping through hoops to avoid regenerating the protos, but I'm curious as to why. Proto generation is fairly fast and relatively rare. If you're worried about the subsequent recompiles of go code then I would suggest generating into a temp directory and comparing the old vs new before replacing.

tamird replied:
This was actually done to make protos depend on external files (like $(PROTOC)) while allowing the use of `$<` - I've just removed `$<` and all this got simpler

petermattis commented on build/protobuf.mk:
```diff
@@ -0,0 +1,68 @@
+
+$(CPP_HEADERS) $(CPP_SOURCES): $(PROTOC) $(CPP_PROTOS)
+	rm -f $(CPP_HEADERS) $(CPP_SOURCES)
+	$(PROTOC) --plugin=$(PROTOC_PLUGIN) -I.:$(GOGOPROTO_PATH) --cpp_out=$(ENGINE_ROOT) $(CPP_PROTOS)
```
I don't think you need the `--plugin=$(PROTOC_PLUGIN)` flag when compiling the cpp protos.

tamird replied:
done

petermattis commented on build/protobuf.mk:
```diff
@@ -0,0 +1,68 @@
+.PHONY:
+protos: $(GO_SOURCES) $(CPP_HEADERS) $(CPP_SOURCES)
+
+$(GO_SOURCES):                 $(PROTOC) $(GO_PROTOS) $(GOGOPROTO_PROTO) $(PROTOC_PLUGIN)
```
Some slightly excessive spacing going on there.

tamird replied:
yeah that was to line it up with the thing below - perhaps unnecessary?

petermattis replied:
Oh, I didn't even notice. Yeah, I think unnecessary.

tamird commented:
I think I've addressed everything. Take another look

petermattis commented:
LGTM

petermattis commented on build/protobuf.mk:
```diff
@@ -0,0 +1,61 @@
+
+$(GO_SOURCES): $(PROTOC) $(GO_PROTOS) $(GOGOPROTO_PROTO) $(PROTOC_PLUGIN)
+	find $(REPO_ROOT) -not -path '*/.*' -name *.pb.go | xargs rm
+	$(PROTOC) -I.:$(GOGOPROTO_PATH) --plugin=$(PROTOC_PLUGIN) --$(PLUGIN_SUFFIX)_out=$(ORG_ROOT) $(GO_PROTOS)
```
You could easily write a wrapper around protoc that looped over the protos to build and invoked protoc once per directory. This would remove the blockage on upstream gogoproto/protoc changes.

tamird replied:
Right, but what happens when our proto packages no longer map to top level directories? :bomb:

petermattis replied:
Not sure I understand. Can you elaborate?

tamird replied:
for instance:
```
├── client # package cockroach.client
│   ├── foo.proto
│   ├── bar.proto
├── storage # package cockroach.storage
│   ├── foo.proto
│   ├── bar.proto
│   ├── engine # package cockroach.storage.engine
│   │   ├── foo.proto
│   │   ├── bar.proto
```
invoking protoc on `storage` will :boom:

bdarnell replied:
I'm sure there's a more efficient way to do this than `xargs -n1 dirname`, but it works (as long as we never have protos in the same directory as a GO_PROTO that are not in GO_PROTOS themselves).

`for d in $(echo $GO_PROTOS | xargs -n1 dirname | sort -u); do $(PROTOC) [other options] $d/*.proto; done`

tschottdorf replied:
Do you mean because it'll bundle the protos from `cockroach.storage{,.engine}` together? You could add `-maxdepth` and (maybe) `-prune` to `find`s arguments to restrict it, and have the loop over all directories which contain at least one proto file at base level, which can also be done with some `find -exec | uniq`.

tamird replied:
OK who wants to write this shell script for me?

tamird committed:
Work around gogo/protobuf#67

Invoke protoc separately for each directory containing protos. h/t
@petermattis.

petermattis commented on build/protobuf.mk:
```diff
@@ -0,0 +1,66 @@
+CPP_HEADERS := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.h)) $(subst $(GOGOPROTO_ROOT),$(ENGINE_ROOT),$(GOGOPROTO_PROTO:%.proto=%.pb.h))
+CPP_SOURCES := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.cc)) $(subst $(GOGOPROTO_ROOT),$(ENGINE_ROOT),$(GOGOPROTO_PROTO:%.proto=%.pb.cc))
+
+.PHONY:
```
Shouldn't this be `.PHONY: protos`? Or does it not matter?

tamird replied:
`cd .. && touch protos && cd - && find . -name *.pb.* | xargs rm && go generate ./...` didn't leave me with missing files, so it seems to work?

petermattis replied:
Hmm. All the `.PHONY` targets in the main makefile do it differently. And the make manual says to do `.PHONY: target`. Let's be consistent.

bdarnell commented on build/protobuf.mk:
```diff
@@ -0,0 +1,66 @@
+$(CPP_HEADERS) $(CPP_SOURCES): $(PROTOC) $(CPP_PROTOS) $(GOGOPROTO_PROTO)
+	find $(REPO_ROOT) -not -path '*/.*' -name *.pb.h -o -name *.pb.cc | xargs rm
+	for dir in $(sort $(dir $(CPP_PROTOS))); do \
+	  $(PROTOC) -I.:$(GOGOPROTO_PATH) --cpp_out=$(ENGINE_ROOT) $(CPP_PROTOS) $$dir/*.proto; \
```
Isn't it redundant to include `$(CPP_PROTOS)` and `$$dir/*.proto` here? Also, `*.proto` will undo the filtering of `CPP_PROTOS` and generate all the protos at once. Can't the cpp generation simply use the old pattern without the for loop?

tamird replied:
ugh. yeah, you're right, i'll revert this bit. this is an ugly wrinkle of this approach if ever the cpp protos need to come from several packages.

petermattis commented:
LGTM modulo the `#include` path question.

tschottdorf commented:
what's the "h/t" in your last commit's message?

tamird commented:
hat tip
